### PR TITLE
User-defined UCR instance IDs

### PR DESCRIPTION
**Pull request by esoergel**
https://docs.google.com/document/d/1RbO8LXfKmFbJPLbshtFNVYSi1hX9nAb_9YPufEPgdVg/edit
Gonna throw this on staging and have someone look at it
:blowfish: :x: :blowfish: 
@emord @orangejenny

### CODE REVIEW

File: corehq/apps/app_manager/static/app_manager/js/modules/report_module.js [hunk 1]
```diff
@@ -318,6 +323,13 @@ hqDefine('app_manager/js/modules/report_module', function () {
                         break;
                     }
                 }
+
+                var duplicatedSlugs = getDuplicatedSlugs();
+                if (duplicatedSlugs.length !== 0) {
+                    alert(gettext("Report codes must be unique.  The following codes are duplicates: ")
+                          + duplicatedSlugs.join(", "));
+                }
```
Comment (esoergel) [inline corehq/apps/app_manager/static/app_manager/js/modules/report_module.js]: I used `alert` here because that's what the above validation check does, but that doesn't actually prevent saving - if the user dismisses the alert, the save goes through.  Is there another pattern for module validation in app builder?

File: corehq/apps/app_manager/suite_xml/features/mobile_ucr.py [hunk 2]
```diff
@@ -449,6 +449,7 @@ def is_valid_mobile_select_filter_type(ui_filter):
     return isinstance(ui_filter, DynamicChoiceListFilter) or isinstance(ui_filter, ChoiceListFilter)
 
 
+@quickcache(['domain'])
```
Comment (emord) [inline corehq/apps/app_manager/suite_xml/features/mobile_ucr.py]: Depending on how confident you are that you're clearing it correctly, you may want to extend this cache beyond 5 minutes.

File: corehq/apps/app_manager/static/app_manager/js/modules/report_module.js (hunk 1 above)
Comment (emord) [reply corehq/apps/app_manager/static/app_manager/js/modules/report_module.js]: I think that generally we allow users to save their app no matter what, even if a build will fail (at least in form builder)

This looks like it only checks instances in the module. Is that a sufficient check? or should it be over app or domain? (We may have talked about this and I forgot the answer)

File: corehq/apps/app_manager/static/app_manager/js/modules/report_module.js [hunk 3]
```diff
@@ -394,6 +401,29 @@ hqDefine('app_manager/js/modules/report_module', function () {
...
+            instanceIdsElsewhere = _.filter(_.keys(uuidsByInstanceId), function (instanceId) {
+                return !_.contains(instanceIdsInThisModule, instanceId);
+            });
+
+        var getDuplicatedSlugs = function () {
+            var instanceIdsInThisModule = _.map(self.reports(), function (r) {return r.instanceId();}),
+                duplicatesWithElsewhere = _.filter(instanceIdsInThisModule, function(iid) {
+                    return _.contains(instanceIdsElsewhere, iid);
+                }),
+                duplicatesHere = _.chain(instanceIdsInThisModule)
```
Comment (orangejenny) [inline corehq/apps/app_manager/static/app_manager/js/modules/report_module.js]: This is a little hard to parse...if `instanceIdsInThisModule` and `instanceIdsElsewhere` are both strings of ids, which is what they look like, couldn't you use `_.intersection(instanceIdsInThisModule, instanceIdsElsewhere)` to return the duplicates with other modules and compare `instanceIdsInThisModule.length` to `_.uniq(instanceIdsInThisModule).length` to see if there are duplicates within this module?

File: corehq/apps/app_manager/static/app_manager/js/modules/report_module.js (hunk 1 above)
Comment (orangejenny) [reply corehq/apps/app_manager/static/app_manager/js/modules/report_module.js]: You can use the `validate_for_build` methods on app manager models to do server-side validation and then add a message to build_errors.html. Grep for "no source module id" to see a simple example of such a validation.

Comment (orangejenny) [reply corehq/apps/app_manager/static/app_manager/js/modules/report_module.js]: Also, for client-side validation, browser alerts aren't a good way to display errors (since the dialog takes control of the whole page, so you can't necessarily see where the error is), but I understand working with the pattern that's already on the page. I'll do a PR to clean it up.

File: corehq/apps/app_manager/static/app_manager/js/modules/report_module.js (hunk 3 above)
Comment (esoergel) [reply corehq/apps/app_manager/static/app_manager/js/modules/report_module.js]: I wanted to display to the user the list of duplicated IDs, not just whether or not IDs were duplicated, which makes the code a little more complex.  I merged in https://github.com/dimagi/commcare-hq/pull/20169 , so I'm switching validation to that method anyways.

**Review by millerdev — APPROVED** [review]
Looks good to me, but I think it would be good to have @orangejenny review/approve of at least the parts she commented on previously as well.

File: corehq/apps/app_manager/static/app_manager/js/modules/report_module.js [hunk 4]
```diff
@@ -396,6 +403,29 @@ hqDefine('app_manager/js/modules/report_module', function () {
...
+        // flag instance ids with uuids outside this module
+        var uuidsByInstanceId = hqImport('hqwebapp/js/initial_page_data').get('uuids_by_instance_id'),
+            uuidsInThisModule = _.pluck(self.reports(), 'uuid'),
+            instanceIdsElsewhere = _.chain(uuidsByInstanceId)
+                .pairs()
+                .filter(function (idPair) { return _.difference(idPair[1], uuidsInThisModule).length; })
+                .map(_.first)
+                .value();
+
+        self.validateSlug = function (instanceId) {
```
[the remark this answers is not in the document]
Comment (esoergel) [reply corehq/apps/app_manager/static/app_manager/js/modules/report_module.js]: `instanceIdsElsewhere` is fixed (although this _could_ query the server to make sure the info is up-to-date), so that should be fine, and the function doesn't use `uuidsInThisModule`.  Are you saying that knockout won't trigger this function at the appropriate times?  Anecdotally, it does seem to work, feel free to play with this test module on staging:
https://staging.commcarehq.org/a/ethan-test/apps/view/ab81ce0c2aaf8df0775ff6712722c72a/module/ae427b9083b972f0bab718e318120c845d1de9b7/
Is there a more "knockout" way to do this?

File: corehq/apps/app_manager/templates/app_manager/partials/mobile_report_configs.html [hunk 5]
```diff
@@ -69,14 +69,20 @@
                             <select class="form-control"
                                     data-bind="options: availableReportIds, optionsText: $root.defaultReportTitle, value: reportId">
                             </select>
-                            <label>{% trans "Report UUID to use in custom expressions:" %}</label>
-                            <span data-bind="text: uuid"></span>
+                            <div data-bind="css: { 'has-error': $root.validateSlug(instanceId()) }">
+                                <label>{% trans "Report code to use in custom expressions:" %}</label>
+                                <span data-bind="visible: !$root.supportCustomUcrSlug, text: uuid"></span>
+                                <input type="text" class="form-inline"
```
Comment (orangejenny) [inline corehq/apps/app_manager/templates/app_manager/partials/mobile_report_configs.html]: Noticed while looking on staging - `form-inline` isn't doing much of anything here. If you use `form-control` on this input and then `form-inline` on the parent div, you'll get the round corners, etc. on the input and the layout should look reasonable.

Comment (esoergel) [reply corehq/apps/app_manager/templates/app_manager/partials/mobile_report_configs.html]: Nice, thanks, just updated it.

**Review by orangejenny — APPROVED** [review]

**Comment by esoergel** [comment]
Sorry stickler, agree to disagree

File: corehq/apps/app_manager/static/app_manager/js/modules/report_module.js [hunk 6]
```diff
@@ -277,7 +281,8 @@ hqDefine('app_manager/js/modules/report_module', function () {
         var currentReports = options.currentReports || [];
         var availableReports = options.availableReports || [];
         var saveURL = options.saveURL;
-        self.supportSyncDelay = options.supportSyncDelay;
+        self.supportSyncDelay = options.mobileUcrVersion !== 1;
+        self.supportCustomUcrSlug = options.mobileUcrVersion >= 2;
```
[the remark this answers is not in the document]
Comment (esoergel) [reply corehq/apps/app_manager/static/app_manager/js/modules/report_module.js]: Ugh, true - know of some JS equivalent to `distutils.version.LooseVersion`?
